sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -125,7 +133,9 @@ case class GroupExpression(children: Seq[Expression]) extends Expression {
  * so that the proper type conversions can be performed in the analyzer.
  */
 trait ExpectsInputTypes {
+  self: Expression =>
 
   def expectedChildTypes: Seq[DataType]
 
+  override def validInputTypes: Boolean = children.map(_.dataType) == expectedChildTypes
```

Comment:
That should be fine right? 

Also - add inline comment explaining this (i.e. things get casted).